### PR TITLE
OmeZarr and OpenOrganelle opening

### PR DESCRIPTION
**Pull request by KateMoreva**

### CODE REVIEW

File: src/main/java/de/embl/cba/mobie/MoBIE.java
Comment (constantinpape) [inline src/main/java/de/embl/cba/mobie/MoBIE.java:72]: It looks like you did some auto formatting on this file. 
In general it would be good not to do this mixed in with other PRs, because it makes reviewing very difficult because there are so many changes. Do you think you could revert the pure formatting changes without too much effort?

**Review by constantinpape — COMMENTED** [review]
Awesome that this works now! But could you revert the auto-formatting changes, @KateMoreva?
This makes finding the actual code changes very difficult.

**Comment by KateMoreva** [comment]
! Problems with opening local ome.zarr files after merging `master`.

**Comment by constantinpape** [comment]
> ! Problems with opening local ome.zarr files after merging `master`.

Ok, let us know if you need help on this. (And in that case please post a stack trace ;))

**Comment by KateMoreva** [comment]
> > ! Problems with opening local ome.zarr files after merging `master`.
> 
> Ok, let us know if you need help on this. (And in that case please post a stack trace ;))

While trying to open ome.zarr from g/kreshuk/pape/Work/mobie/covid-if-project I get 

Exception in thread "main" java.lang.RuntimeException: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_ARRAY but was END_ARRAY at path $.sources[0][1]
	at de.embl.cba.mobie.serialize.JsonHelper.createObjectFromJsonValue(JsonHelper.java:56)
	at de.embl.cba.mobie.serialize.SourceTransformerListAdapter.deserialize(SourceTransformerListAdapter.java:35)
	at de.embl.cba.mobie.serialize.SourceTransformerListAdapter.deserialize(SourceTransformerListAdapter.java:13)
	at com.google.gson.internal.bind.TreeTypeAdapter.read(TreeTypeAdapter.java:69)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:187)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	at com.google.gson.Gson.fromJson(Gson.java:897)
	at com.google.gson.Gson.fromJson(Gson.java:846)
	at de.embl.cba.mobie.serialize.DatasetJsonParser.parseDataset(DatasetJsonParser.java:26)
	at de.embl.cba.mobie.MoBIE.openDataset(MoBIE.java:183)
	at de.embl.cba.mobie.MoBIE.openDataset(MoBIE.java:122)
	at de.embl.cba.mobie.MoBIE.<init>(MoBIE.java:91)
	at projects.OpenLocalOmeZarr.main(OpenLocalOmeZarr.java:14)
Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_ARRAY but was END_ARRAY at path $.sources[0][1]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:226)
	at com.google.gson.Gson.fromJson(Gson.java:932)
	at com.google.gson.Gson.fromJson(Gson.java:1003)
	at com.google.gson.internal.bind.TreeTypeAdapter$GsonContextImpl.deserialize(TreeTypeAdapter.java:162)
	at de.embl.cba.mobie.serialize.JsonHelper.createObjectFromJsonValue(JsonHelper.java:51)
	... 18 more
Caused by: java.lang.IllegalStateException: Expected BEGIN_ARRAY but was END_ARRAY at path $.sources[0][1]
	at com.google.gson.internal.bind.JsonTreeReader.expect(JsonTreeReader.java:163)
	at com.google.gson.internal.bind.JsonTreeReader.beginArray(JsonTreeReader.java:72)
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:80)
	at com.google.gson.internal.bind.CollectionTypeAdapterFactory$Adapter.read(CollectionTypeAdapterFactory.java:61)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:174)
	at com.google.gson.internal.bind.MapTypeAdapterFactory$Adapter.read(MapTypeAdapterFactory.java:145)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:131)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:222)
	... 22 more

**Comment by constantinpape** [comment]
Ok, @tischi could you have a look at the error?

**Comment by tischi** [comment]
@constantinpape did you update the JSON for the `gridTransform` to the latest spec already in this project?

**Comment by constantinpape** [comment]
> @constantinpape did you update the JSON for the `gridTransform` to the latest spec already in this project?

Good point, I haven't updated it yet. Will do it and this should fix the issue.

**Comment by constantinpape** [comment]
@KateMoreva I updated the project, could you please try again?

**Comment by KateMoreva** [comment]
> @KateMoreva I updated the project, could you please try again?

Thank you, this error is now solved, but I have a different one now :)

**Comment by constantinpape** [comment]
Ok, let us know if we can help fixing it ;)

**Comment by KateMoreva** [comment]
> Ok, let us know if we can help fixing it ;)

Everything works fine on my side.

File: src/main/java/de/embl/cba/mobie/MoBIE.java
Comment (constantinpape) [inline src/main/java/de/embl/cba/mobie/MoBIE.java:279]: Could you fix the indentation here? (BdvN5 is misaligned)

File: src/main/java/de/embl/cba/mobie/MoBIE.java
Comment (constantinpape) [inline src/main/java/de/embl/cba/mobie/MoBIE.java:282]: Shouldn't there also be a case for `OmeZarrS3`?

File: src/main/java/de/embl/cba/mobie/MoBIE.java
Comment (constantinpape) [inline src/main/java/de/embl/cba/mobie/MoBIE.java:285]: Shouldn't there also be a case for BdvOmeZarrS3?

**Review by constantinpape — COMMENTED** [review]
Looks good to me, but I think it's missing cases for `OmeZarrS3` and `BdvOmeZarrS3`.
(Note that we don't have example data for `OmeZarrS3` yet, I will try to upload this soon).

**Comment by constantinpape** [comment]
@tischi could you also have a look at the changes please?

**Comment by KateMoreva** [comment]
> Looks good to me, but I think it's missing cases for `OmeZarrS3` and `BdvOmeZarrS3`.
> (Note that we don't have example data for `OmeZarrS3` yet, I will try to upload this soon).

Well, I did not notice that a different data type was specified in the s3 version.

**Comment by constantinpape** [comment]
> > 
> 
> Well, I did not notice that a different data type was specified in the s3 version.

Yes, we use a different type for the file system and s3 version to differentiate them. You can find an example for this here:
`/g/kreshuk/pape/Work/mobie/covid-em-datasets/ngff-example/data/Covid19-S4-Area2/dataset.json`

**Comment by KateMoreva** [comment]
> > > 
> > 
> > 
> > Well, I did not notice that a different data type was specified in the s3 version.
> 
> Yes, we use a different type for the file system and s3 version to differentiate them. You can find an example for this here:
> `/g/kreshuk/pape/Work/mobie/covid-em-datasets/ngff-example/data/Covid19-S4-Area2/dataset.json`

yes, I used this one and the `https://s3.embl.de/i2k-2020/project-bdv-ome-zarr` for bdv.ome.zarr format opening testing.

**Comment by constantinpape** [comment]
> yes, I used this one and the `https://s3.embl.de/i2k-2020/project-bdv-ome-zarr` for bdv.ome.zarr format opening testing.

Ok, if you look in the json you will see that it lists both `bdv.ome.zarr` as well as `bdv.ome.zarr.s3` as fileformats. I would think that it only works from s3 if you have the latter. (But it sounds like it works with the current code, which I don't quite understand).

**Comment by KateMoreva** [comment]
> > yes, I used this one and the `https://s3.embl.de/i2k-2020/project-bdv-ome-zarr` for bdv.ome.zarr format opening testing.
> 
> Ok, if you look in the json you will see that it lists both `bdv.ome.zarr` as well as `bdv.ome.zarr.s3` as fileformats. I would think that it only works from s3 if you have the latter. (But it sounds like it works with the current code, which I don't quite understand).

It seems that in s3 version there are 2 file formats specified: 
```
"imageDataFormats": [
    "bdv.ome.zarr",
    "bdv.ome.zarr.s3"
  ],
```

that is why the code is working for s3 and non-s3 versions.

**Comment by KateMoreva** [comment]
@constantinpape  is there possibly an s3 variant with only one imageDataFormat in JSON?

**Comment by constantinpape** [comment]
> @constantinpape is there possibly an s3 variant with only one imageDataFormat in JSON?

No, there is only the one with both; and this is also how it will be for most of the actual projects (because we first add the local data, corresponding to e.g. `ome.zarr` and then the s3 data, corresponding to `ome.zarr.s3`).
There should be some mechanism that selects the correct one of the two formats depending on where the project is loaded from. @tischi can you explain the details here?

**Comment by KateMoreva** [comment]
> > @constantinpape is there possibly an s3 variant with only one imageDataFormat in JSON?
> 
> No, there is only the one with both; and this is also how it will be for most of the actual projects (because we first add the local data, corresponding to e.g. `ome.zarr` and then the s3 data, corresponding to `ome.zarr.s3`).
> There should be some mechanism that selects the correct one of the two formats depending on where the project is loaded from. @tischi can you explain the details here?

Ok, then I'll add the s3 variant, but I do not know how to test if the reader correctly opens s3 part separately.

**Comment by tischi** [comment]
> There should be some mechanism that selects the correct one of the two formats depending on where the project is loaded from. @tischi can you explain the details here?

Currently this is not automated yet, but you have to specify it, e.g.

```
new MoBIE("/Volumes/emcf/pape/mobie-test-projects/", MoBIESettings.settings().imageDataFormat( ImageDataFormat.BdvN5 ));
```

But for the Fiji UI, we could implement that it guesses the format from the entry point, i.e. the path to the project (picking the one from the project.json that matches best). Or presents the user with a UI to choose.

**Review by constantinpape — APPROVED** [review]
This looks good to me know.
(I think there is still some things missing for "ome.zarr.s3", but we also don't have a complete example project for this yet, so we can leave it to a different PR).
Let's wait till @tischi also had some time to look at this before merging.

File: src/main/java/de/embl/cba/mobie/MoBIE.java
Comment (KateMoreva) [reply src/main/java/de/embl/cba/mobie/MoBIE.java]: Sorry for autoformatting. It will not happen again.

**Comment by KateMoreva** [comment]
> This looks good to me know.
> (I think there is still some things missing for "ome.zarr.s3", but we also don't have a complete example project for this yet, so we can leave it to a different PR).
> Let's wait till @tischi also had some time to look at this before merging.

Thank you, yes it lacks `ome.zarr.s3`, but I do not want to add the format without testing.

**Comment by tischi** [comment]
Regarding the formatting: I am using here the imglib2-style formatting, with lots of spaces 😄 .
Indeed it's a bit hard to review now, but no worries. I'll just merge as I have the feeling the actual changes should not be conflicting with the existing functionality! Thanks for all the work!!

**Pull request merged**
